File: sdk/dotnet/MockForge.Sdk/MockServerException.cs
namespace MockForge.Sdk;

using System.Collections.Generic;

/// <summary>
/// Standardized exception class for MockServer operations
/// </summary>
public class MockServerException : Exception
{
    /// <summary>
    /// Error codes for MockServer operations
    /// </summary>
    public enum ErrorCode
    {
        CliNotFound,
        ServerStartFailed,
        PortDetectionFailed,
        AdminApiError,
        HealthCheckTimeout,
        InvalidConfig,
        StubNotFound,
        NetworkError,
        UnknownError
    }

    /// <summary>
    /// Error code for this exception
    /// </summary>
    public ErrorCode Code { get; }

    /// <summary>
    /// Additional error details
    /// </summary>
    public Dictionary<string, object> Details { get; }

    public MockServerException(ErrorCode code, string message) : base(message)
    {
        Code = code;
        Details = new Dictionary<string, object>();
    }

    public MockServerException(ErrorCode code, string message, Exception innerException)
        : base(message, innerException)
    {
        Code = code;
        Details = new Dictionary<string, object>();
    }

    /// <summary>
    /// Create an error for CLI not found
    /// </summary>
    public static MockServerException CliNotFound(Exception? cause = null)
    {
        var ex = new MockServerException(
            ErrorCode.CliNotFound,
            "MockForge CLI not found. Install with: cargo install mockforge-cli",
            cause
        );
        ex.Details["hint"] = "Ensure mockforge is in your PATH";
        return ex;
    }

    /// <summary>
    /// Create an error for server start failure
    /// </summary>
    public static MockServerException ServerStartFailed(string message, Exception? cause = null)
    {
        return new MockServerException(
            ErrorCode.ServerStartFailed,
            $"Failed to start MockForge server: {message}",
            cause
        );
    }

    /// <summary>
    /// Create an error for port detection failure
    /// </summary>
    public static MockServerException PortDetectionFailed(Exception? cause = null)
    {
        var ex = new MockServerException(
            ErrorCode.PortDetectionFailed,
            "Failed to detect server port from MockForge output. The server may have failed to start.",
            cause
        );
        ex.Details["hint"] = "Check that mockforge CLI is installed and the server started successfully";
        return ex;
    }

    /// <summary>
    /// Create an error for Admin API operations
    /// </summary>
    public static MockServerException AdminApiError(string operation, string message, Exception? cause = null)
    {
        var ex = new MockServerException(
            ErrorCode.AdminApiError,
            $"Admin API {operation} failed: {message}",
            cause
        );
        ex.Details["operation"] = operation;
        return ex;
    }

    /// <summary>
    /// Create an error for health check timeout
    /// </summary>
    public static MockServerException HealthCheckTimeout(int timeout, int port)
    {
        var ex = new MockServerException(
            ErrorCode.HealthCheckTimeout,
            $"Health check timed out after {timeout}ms. Could not connect to http://127.0.0.1:{port}/health"
        );
        ex.Details["timeout"] = timeout;
        ex.Details["port"] = port;
        ex.Details["hint"] = "Check that the server started successfully";
        return ex;
    }

    /// <summary>
    /// Create an error for invalid configuration
    /// </summary>
    public static MockServerException InvalidConfig(string message)
    {
        return new MockServerException(
            ErrorCode.InvalidConfig,
            $"Invalid configuration: {message}"
        );
    }

    /// <summary>
    /// Create an error for stub not found
    /// </summary>
    public static MockServerException StubNotFound(string method, string path)
    {
        var ex = new MockServerException(
            ErrorCode.StubNotFound,
            $"Stub not found: {method} {path}"
        );
        ex.Details["method"] = method;
        ex.Details["path"] = path;
        return ex;
    }

    /// <summary>
    /// Create an error for network operations
    /// </summary>
    public static MockServerException NetworkError(string message, Exception? cause = null)
    {
        return new MockServerException(
            ErrorCode.NetworkError,
            $"Network error: {message}",
            cause
        );
    }
}

File: sdk/java/src/main/java/com/mockforge/sdk/MockServerException.java
package com.mockforge.sdk;

import java.util.HashMap;
import java.util.Map;

/**
 * Standardized exception class for MockServer operations
 */
public class MockServerException extends Exception {
    /**
     * Error codes for MockServer operations
     */
    public enum ErrorCode {
        CLI_NOT_FOUND,
        SERVER_START_FAILED,
        PORT_DETECTION_FAILED,
        ADMIN_API_ERROR,
        HEALTH_CHECK_TIMEOUT,
        INVALID_CONFIG,
        STUB_NOT_FOUND,
        NETWORK_ERROR,
        UNKNOWN_ERROR
    }

    private final ErrorCode code;
    private final Map<String, Object> details;

    public MockServerException(ErrorCode code, String message) {
        super(message);
        this.code = code;
        this.details = new HashMap<>();
    }

    public MockServerException(ErrorCode code, String message, Throwable cause) {
        super(message, cause);
        this.code = code;
        this.details = new HashMap<>();
    }

    public ErrorCode getCode() {
        return code;
    }

    public Map<String, Object> getDetails() {
        return details;
    }

    /**
     * Create an error for CLI not found
     */
    public static MockServerException cliNotFound(Throwable cause) {
        MockServerException e = new MockServerException(
            ErrorCode.CLI_NOT_FOUND,
            "MockForge CLI not found. Install with: cargo install mockforge-cli",
            cause
        );
        e.details.put("hint", "Ensure mockforge is in your PATH");
        return e;
    }

    /**
     * Create an error for server start failure
     */
    public static MockServerException serverStartFailed(String message, Throwable cause) {
        return new MockServerException(
            ErrorCode.SERVER_START_FAILED,
            "Failed to start MockForge server: " + message,
            cause
        );
    }

    /**
     * Create an error for port detection failure
     */
    public static MockServerException portDetectionFailed(Throwable cause) {
        MockServerException e = new MockServerException(
            ErrorCode.PORT_DETECTION_FAILED,
            "Failed to detect server port from MockForge output. The server may have failed to start.",
            cause
        );
        e.details.put("hint", "Check that mockforge CLI is installed and the server started successfully");
        return e;
    }

    /**
     * Create an error for Admin API operations
     */
    public static MockServerException adminApiError(String operation, String message, Throwable cause) {
        MockServerException e = new MockServerException(
            ErrorCode.ADMIN_API_ERROR,
            "Admin API " + operation + " failed: " + message,
            cause
        );
        e.details.put("operation", operation);
        return e;
    }

    /**
     * Create an error for health check timeout
     */
    public static MockServerException healthCheckTimeout(int timeout, int port) {
        MockServerException e = new MockServerException(
            ErrorCode.HEALTH_CHECK_TIMEOUT,
            String.format("Health check timed out after %dms. Could not connect to http://127.0.0.1:%d/health", timeout, port)
        );
        e.details.put("timeout", timeout);
        e.details.put("port", port);
        e.details.put("hint", "Check that the server started successfully");
        return e;
    }

    /**
     * Create an error for invalid configuration
     */
    public static MockServerException invalidConfig(String message) {
        return new MockServerException(
            ErrorCode.INVALID_CONFIG,
            "Invalid configuration: " + message
        );
    }

    /**
     * Create an error for stub not found
     */
    public static MockServerException stubNotFound(String method, String path) {
        MockServerException e = new MockServerException(
            ErrorCode.STUB_NOT_FOUND,
            "Stub not found: " + method + " " + path
        );
        e.details.put("method", method);
        e.details.put("path", path);
        return e;
    }

    /**
     * Create an error for network operations
     */
    public static MockServerException networkError(String message, Throwable cause) {
        return new MockServerException(
            ErrorCode.NETWORK_ERROR,
            "Network error: " + message,
            cause
        );
    }
}
